File: backend/services/avatar_engine.py
from typing import Dict, List, Optional
import time

SEBI_FOOTER = (
    "REGULATORY DISCLOSURE (SEBI): Investments in securities market are subject to market risks. "
    "Read all scheme-related documents carefully. Avatar Base is an AI system — not a SEBI-registered IA."
)
TAX_DISCLAIMER = (
    "This is general tax-planning information. Consult a Chartered Accountant for final tax filings."
)

SEGMENT_TONES = {
    "CHILD": "encouraging",
    "STUDENT": "aspirational",
    "YOUNG_EMPLOYED": "efficient",
    "SENIOR": "reassuring"
}

def generate_micro_copy_card(card_type: str, segment: str, data: dict) -> dict:
    """
    Generates a structured 3-part card: What (Headline) -> Why (Reasoning) -> Action (CTA).
    Applies segment-specific tones and legal compliance footers.
    """
    tone = SEGMENT_TONES.get(segment.upper(), "efficient")
    card_id = f"card-{card_type.lower()}-{int(time.time())}"
    
    # Defaults
    accent = "blue"
    type_badge = "Advisory"
    what = ""
    why = ""
    action_label = "View Details"
    action_url = "/dashboard"
    auth_level = 1
    disclosure = SEBI_FOOTER
    data_used = []
    confidence = 0.90
    is_time_sensitive = False
    
    if card_type == "TAX_80C":
        accent = "tax"
        type_badge = "Tax Savings"
        remaining = data.get("remaining", 0)
        savings = int(remaining * 0.30)
        instrument = data.get("instrument", "ELSS Mutual Fund")
        
        if tone == "efficient":
            what = f"Invest ₹{remaining:,} before March 31 — save up to ₹{savings:,} in tax."
            why = f"You have ₹{remaining:,} left in your Section 80C limit this year. Deploying into an {instrument} lowers your taxable income directly."
        elif tone == "reassuring":
            what = f"₹{remaining:,} still available to save on tax — let's put it to work safely."
            why = f"You have ₹{remaining:,} left of your annual tax-saving limit. A safe {instrument} can secure this rebate for you before the March 31 deadline."
        elif tone == "aspirational":
            what = f"Save ₹{savings:,} on tax and build future wealth with ₹{remaining:,} in 80C."
            why = f"Cutting your tax bill by investing ₹{remaining:,} in an {instrument} creates a disciplined compound growth asset for your future goals."
        else:
            what = f"Grow your savings! ₹{remaining:,} can be saved tax-free."
            why = f"Your parents can invest this ₹{remaining:,} in a kids savings growth fund, saving on taxes and planning for your future goals!"
            
        action_label = "Invest Now"
        action_url = f"/invest/80c?amount={remaining}"
        auth_level = 2
        disclosure = TAX_DISCLAIMER
        data_used = ["TAX_PROFILE", "TRANSACTION_HISTORY"]
        confidence = 0.92
        is_time_sensitive = True
        
    elif card_type == "PORTFOLIO_DRIFT":
        accent = "portfolio"
        type_badge = "Portfolio Alert"
        asset_class = data.get("asset_class", "EQUITY")
        drift = data.get("drift", 0.0)
        direction = data.get("direction", "OVERWEIGHT")
        curr = data.get("current", 0.0)
        target = data.get("target", 0.0)
        
        what = f"Your {asset_class} is {direction.lower()} by {abs(drift):.1f}% — rebalancing recommended."
        why = f"Market movements shifted your allocation. {asset_class} is currently {curr}% vs your target of {target}%. Rebalancing helps lock in gains and control risk."
        action_label = "Review Rebalancing Options"
        action_url = f"/portfolio/rebalance"
        auth_level = 2
        disclosure = SEBI_FOOTER
        data_used = ["PORTFOLIO_DATA", "MARKET_DATA"]
        confidence = 0.95
        
    elif card_type == "SALARY_SPLIT":
        accent = "salary"
        type_badge = "Salary Day Splitter"
        net_salary = data.get("net_salary", 0)
        elss_sip = data.get("elss_sip", 0)
        nps_contrib = data.get("nps_contrib", 0)
        ef_contrib = data.get("ef_contrib", 0)
        tax_saved = data.get("tax_saving", 0)
        
        what = f"Your salary of ₹{net_salary:,} is credited. Here is the smart split."
        why = f"We suggest allocating ₹{ef_contrib:,} to your emergency fund, ₹{elss_sip:,} to your ELSS SIP, and ₹{nps_contrib:,} to NPS. This split saves an estimated ₹{tax_saved:,} in taxes."
        action_label = "Confirm Split & Invest"
        action_url = "/salary-split/confirm"
        auth_level = 2
        disclosure = SEBI_FOOTER + " " + TAX_DISCLAIMER
        data_used = ["TRANSACTION_HISTORY", "TAX_PROFILE", "PORTFOLIO_DATA"]
        confidence = 0.91
        is_time_sensitive = True
        
    elif card_type == "FRAUD_ALERT":
        accent = "fraud"
        type_badge = "Security Alert"
        desc = data.get("description", "We detected unusual transactions.")
        risk = int(data.get("risk_score", 0.8) * 100)
        
        what = "Unusual activity detected on your account — block recommended."
        why = f"{desc} Risk score: {risk}%. If this wasn't you, block your debit card and account immediately to prevent unauthorized outflows."
        action_label = "Block Account Now"
        action_url = "/emergency/block-account"
        auth_level = 1
        disclosure = "This is a real-time security alert from Avatar Base Fraud Sentinel. Contact bank support: 1800-XXX-XXXX."
        data_used = ["TRANSACTION_HISTORY", "DEVICE_FINGERPRINT", "BEHAVIORAL_BIOMETRICS"]
        confidence = data.get("risk_score", 0.85)
        is_time_sensitive = True
        
    elif card_type == "WHS_IMPROVEMENT":
        accent = "whs"
        type_badge = "Wealth Health"
        score = data.get("score", 75)
        grade = data.get("grade", "B")
        top_action = data.get("top_action", "Build Emergency Reserve")
        
        what = f"Your Wealth Health Score is {score}/100 — Grade {grade}"
        why = f"Your biggest opportunity to boost your score is: '{top_action}'. Strengthening this pillar raises your long-term security and score."
        action_label = top_action
        action_url = "/plan/whs-improvement"
        auth_level = 1
        disclosure = "WHS is an illustrative dashboard index indicator, not a yield guarantee."
        data_used = ["TRANSACTION_HISTORY", "PORTFOLIO_DATA", "INSURANCE_DATA"]
        confidence = 0.90
        
    elif card_type == "SENIOR_80TTB":
        accent = "tax"
        type_badge = "Section 80TTB"
        fd_interest = data.get("fd_interest", 0)
        eligible = min(fd_interest, 50000)
        
        what = f"Claim up to ₹{eligible:,} in tax-free FD interest under Section 80TTB."
        why = f"You have earned ₹{fd_interest:,} in interest. Under Section 80TTB, senior citizens can deduct up to ₹50,000 of interest income from their tax liabilities."
        action_label = "Optimize FDs"
        action_url = "/tax/80ttb"
        auth_level = 2
        disclosure = TAX_DISCLAIMER
        data_used = ["TAX_PROFILE", "TRANSACTION_HISTORY"]
        confidence = 0.95

    return {
        "cardId": card_id,
        "type": card_type,
        "segment": segment,
        "accent": accent,
        "typeBadge": type_badge,
        "what": what,
        "why": why,
        "actionLabel": action_label,
        "actionUrl": action_url,
        "authLevel": auth_level,
        "disclosure": disclosure,
        "dataUsed": data_used,
        "explainRef": f"/why/{card_id}",
        "confidence": confidence,
        "isTimeSensitive": is_time_sensitive,
        "createdAt": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime())
    }

def generate_message(segment: str) -> str:
    """Backward compatibility greeting route."""
    messages = {
        "children": "Hey! Did you know saving ₹10 a day buys a huge toy at the end of the year?",
        "student": "Welcome to your financial journey. Nova is ready. First step: Build an emergency fund.",
        "young_employed": "Salary credited! Arya suggests allocating 20% to your investments.",
        "senior": "Namaste. Sahaya is monitoring your account. We blocked 2 suspicious IPs today."
    }
    return messages.get(segment.lower(), "Welcome to Avatar Base AI.")

File: frontend/js/avatar-engine.js
/**
 * Avatar Base — Avatar NLG Engine (Frontend)
 * Generates 3-part micro-copy cards: What → Why → Action
 * Tone-adaptive per lifecycle segment
 */

import { fmt } from './tax-optimizer.js';

const SEGMENT_TONES = {
  CHILD:          'encouraging',
  STUDENT:        'aspirational',
  YOUNG_EMPLOYED: 'efficient',
  SENIOR:         'reassuring',
};

const SEBI_FOOTER = `⚠️ Investments subject to market risk. Read all documents carefully. Avatar Base is an AI system, not a SEBI-registered human advisor. All equity suggestions are sourced from SEBI-registered Research Analysts only.`;
const TAX_DISCLAIMER = `This is general tax-planning information, not licensed tax advice. Consult a Chartered Accountant for personalized tax guidance.`;

// ── Avatar Personas ───────────────────────────────────────────────────
const AVATAR_PERSONAS = {
  YOUNG_EMPLOYED: { name: 'Arya', greeting: "Let's make your money work harder." },
  STUDENT:        { name: 'Nova', greeting: "Building your future, one step at a time." },
  SENIOR:         { name: 'Sahaya', greeting: "I'm here to help you stay safe and comfortable." },
  CHILD:          { name: 'Buddy', greeting: "Let's save up for something awesome!" },
};

// ── Micro-Copy Card Generator ──────────────────────────────────────────
export function generateCard({ type, segment, data, modelVersion = 'v1.2.0' }) {
  const tone = SEGMENT_TONES[segment] || 'efficient';
  const cardId = `card-${type}-${Date.now()}`;

  const generators = {
    TAX_80C:       () => gen80CCard(segment, tone, data, cardId),
    PORTFOLIO_DRIFT: () => genDriftCard(segment, tone, data, cardId),
    SALARY_SPLIT:  () => genSalarySplitCard(segment, tone, data, cardId),
    FRAUD_ALERT:   () => genFraudCard(segment, data, cardId),
    WHS_IMPROVEMENT: () => genWHSCard(segment, tone, data, cardId),
    CAPITAL_GAINS: () => genCGCard(segment, tone, data, cardId),
    REGIME_SWITCH: () => genRegimeCard(segment, tone, data, cardId),
    INSURANCE_GAP: () => genInsuranceCard(segment, tone, data, cardId),
    NPS_PUSH:      () => genNPSCard(segment, tone, data, cardId),
    SENIOR_80TTB:  () => genSenior80TTBCard(segment, data, cardId),
  };

  const gen = generators[type];
  if (!gen) return null;
  return { ...gen(), cardId, type, segment, modelVersion, createdAt: new Date().toISOString() };
}

// ── Card Templates ─────────────────────────────────────────────────────
function gen80CCard(segment, tone, data, cardId) {
  const remaining = data.remaining || 0;
  const savings = Math.round(remaining * 0.30);
  const instrument = data.topRecommendation || 'ELSS fund';
  const month = new Date().toLocaleString('default', { month: 'long' });

  const whats = {
    efficient:   `Invest ₹${fmt(remaining)} before March 31 — save up to ₹${fmt(savings)} in tax.`,
    reassuring:  `₹${fmt(remaining)} left in your 80C limit — let's make sure it works for you.`,
    aspirational:`Your 80C limit has ₹${fmt(remaining)} unused — start investing in your future today.`,
    encouraging: `Help grow your savings! ₹${fmt(remaining)} is still available to invest.`,
  };

  const whys = {
    efficient:   `The government allows up to ₹1.5L tax-free investments each year. Your ${instrument} investment reduces your taxable income directly. ${month} is a great time to deploy this.`,
    reassuring:  `This is a safe way to reduce your tax bill this year. An ${instrument} with lock-in of 3 years offers good returns with government backing on tax benefits.`,
    aspirational:`Every rupee you invest in 80C now cuts your tax bill AND builds your future wealth. An ${instrument} grows at market rates while protecting your income.`,
    encouraging: `Your parents have set aside ₹${fmt(remaining)} to grow in a savings plan. It's like a magic jar that grows bigger every year!`,
  };

  return {
    accent: 'tax',
    typeBadge: 'Tax Savings',
    what: whats[tone] || whats.efficient,
    why: whys[tone] || whys.efficient,
    actionLabel: 'Invest Now',
    actionUrl: `/invest/80c?amount=${remaining}`,
    authLevel: 2,
    disclosure: TAX_DISCLAIMER,
    dataUsed: ['TAX_PROFILE', 'TRANSACTION_HISTORY'],
    explainRef: `/why/${cardId}`,
    confidence: 0.92,
    isTimeSensitive: true,
    urgency: data.urgency || 'MEDIUM',
  };
}

function genDriftCard(segment, tone, data, cardId) {
  const { assetClass, drift, direction, current, target } = data;
  const absDrift = Math.abs(drift).toFixed(1);
  return {
    accent: 'portfolio',
    typeBadge: 'Portfolio Alert',
    what: `Your ${assetClass} is ${direction.toLowerCase()} by ${absDrift}% — rebalancing recommended.`,
    why: `Markets shifted your portfolio away from your target. ${assetClass} is now ${current}% (target: ${target}%). Keeping allocation in check protects you from outsized risk in one area.`,
    actionLabel: 'Review Rebalancing',
    actionUrl: `/portfolio/rebalance?alert_id=${cardId}`,
    authLevel: 2,
    disclosure: SEBI_FOOTER,
    dataUsed: ['PORTFOLIO_DATA', 'MARKET_DATA'],
    explainRef: `/why/${cardId}`,
    confidence: 0.95,
    isTimeSensitive: false,
  };
}

function genSalarySplitCard(segment, tone, data, cardId) {
  const { netSalary, elssSIP, npsContrib, efContrib, taxSaving } = data;
  return {
    accent: 'salary',
    typeBadge: 'Salary Day',
    what: `Your ₹${fmt(netSalary)} salary is in — here's your smart money split.`,
    why: `₹${fmt(efContrib)} to emergency fund, ₹${fmt(elssSIP)} as ELSS tax-saver SIP, ₹${fmt(npsContrib)} to NPS. This split could save you ₹${fmt(taxSaving)} in taxes this year — all in one tap.`,
    actionLabel: 'Confirm & Activate SIPs',
    actionUrl: '/salary-split/confirm',
    authLevel: 2,
    disclosure: SEBI_FOOTER,
    dataUsed: ['TRANSACTION_HISTORY', 'TAX_PROFILE', 'PORTFOLIO_DATA'],
    explainRef: `/why/${cardId}`,
    confidence: 0.91,
    isTimeSensitive: true,
  };
}

function genFraudCard(segment, data, cardId) {
  return {
    accent: 'fraud',
    typeBadge: 'Security Alert',
    what: `Unusual activity detected on your account — action required.`,
    why: `${data.description || 'We noticed a transaction that doesn\'t match your normal pattern'}. Risk score: ${((data.riskScore || 0.8) * 100).toFixed(0)}%. If this was not you, block your account immediately.`,
    actionLabel: 'Block Account & Verify',
    actionUrl: '/emergency/block-account',
    authLevel: 1,
    disclosure: 'This is a real-time security alert from Avatar Base Fraud Sentinel. Contact the bank helpline: 1800-XXX-XXXX.',
    dataUsed: ['TRANSACTION_HISTORY', 'DEVICE_FINGERPRINT', 'BEHAVIORAL_BIOMETRICS'],
    explainRef: `/why/${cardId}`,
    confidence: data.riskScore || 0.85,
    isTimeSensitive: true,
    isDismissible: false,
  };
}

function genWHSCard(segment, tone, data, cardId) {
  const { score, grade, gradeColor, topAction } = data;
  const tonePrefix = {
    efficient: "Your financial health at a glance:",
    reassuring: "Here's how your finances are looking:",
    aspirational: "Your wealth journey score:",
    encouraging: "Let's see how well you're saving!",
  };
  return {
    accent: 'whs',
    typeBadge: 'Wealth Health',
    what: `${tonePrefix[tone]} Score ${score}/100 — Grade ${grade}`,
    why: data.explanation || `${topAction?.label || 'Review your financial profile'} is your biggest opportunity to improve your score this month.`,
    actionLabel: topAction?.label || 'See Full Analysis',
    actionUrl: topAction?.url || '/dashboard',
    authLevel: 1,
    disclosure: 'Wealth Health Score is an illustrative indicator, not a guarantee of financial outcomes.',
    dataUsed: ['TRANSACTION_HISTORY', 'PORTFOLIO_DATA', 'INSURANCE_DATA'],
    explainRef: `/why/${cardId}`,
    confidence: 0.90,
    isTimeSensitive: false,
  };
}

function genCGCard(segment, tone, data, cardId) {
  const { totalTax, daysToFYEnd, totalHarvestSaving, netLTCG, netSTCG } = data;
  return {
    accent: 'tax',
    typeBadge: '📋 Capital Gains',
    what: `Estimated capital gains tax: ₹${fmt(totalTax)} — ${daysToFYEnd} days to act.`,
    why: `₹${fmt(netLTCG)} in long-term and ₹${fmt(netSTCG)} in short-term gains this year. Booking strategic losses could save up to ₹${fmt(totalHarvestSaving)}. March 31 is the last date.`,
    actionLabel: 'Review Tax-Loss Harvesting',
    actionUrl: `/tax/capital-gains`,
    authLevel: 2,
    disclosure: TAX_DISCLAIMER + ' ' + SEBI_FOOTER,
    dataUsed: ['PORTFOLIO_DATA', 'CAPITAL_GAINS_LEDGER'],
    explainRef: `/why/${cardId}`,
    confidence: 0.89,
    isTimeSensitive: true,
  };
}

function genRegimeCard(segment, tone, data, cardId) {
  const { recommended, savings, verdict } = data;
  return {
    accent: 'tax',
    typeBadge: '🔄 Tax Regime',
    what: `${recommended} Tax Regime is better for you — saves ₹${fmt(savings)} this year.`,
    why: verdict,
    actionLabel: `Switch to ${recommended} Regime`,
    actionUrl: '/tax/regime-switch',
    authLevel: 3,
    disclosure: TAX_DISCLAIMER,
    dataUsed: ['TAX_PROFILE', 'TRANSACTION_HISTORY'],
    explainRef: `/why/${cardId}`,
    confidence: 0.88,
    isTimeSensitive: true,
  };
}

function genInsuranceCard(segment, tone, data, cardId) {
  return {
    accent: 'insurance',
    typeBadge: 'Insurance Gap',
    what: `Your ${data.gapType || 'health'} insurance coverage has a gap of ₹${fmt(data.gap || 0)}.`,
    why: `Your current coverage of ₹${fmt(data.current || 0)} is below the recommended ₹${fmt(data.recommended || 0)} for your family size. Adequate coverage protects your savings from medical emergencies.`,
    actionLabel: 'Explore Insurance Options',
    actionUrl: '/protect/insurance',
    authLevel: 2,
    disclosure: 'Insurance is subject to terms and conditions. Read policy documents carefully.',
    dataUsed: ['INSURANCE_DATA'],
    explainRef: `/why/${cardId}`,
    confidence: 0.87,
    isTimeSensitive: false,
  };
}

function genNPSCard(segment, tone, data, cardId) {
  const headroom = data.remaining || 50000;
  return {
    accent: 'tax',
    typeBadge: 'NPS Opportunity',
    what: `Save ₹${fmt(Math.round(headroom * 0.30))} more in tax with NPS — ₹${fmt(headroom)} headroom left.`,
    why: `Section 80CCD(1B) gives you an additional ₹50,000 deduction beyond your 80C limit — exclusive to NPS. Even in the ${data.taxBracket || '30%'} bracket, this saves ₹${fmt(Math.round(headroom * 0.30))} annually.`,
    actionLabel: 'Contribute to NPS',
    actionUrl: '/invest/nps',
    authLevel: 2,
    disclosure: TAX_DISCLAIMER,
    dataUsed: ['TAX_PROFILE'],
    explainRef: `/why/${cardId}`,
    confidence: 0.93,
    isTimeSensitive: true,
  };
}

function genSenior80TTBCard(segment, data, cardId) {
  const fdInterest = data.fdInterest || 0;
  const eligible = Math.min(fdInterest, 50000);
  return {
    accent: 'tax',
    typeBadge: 'Section 80TTB',
    what: `As a senior, your FD interest up to ₹50,000 is tax-deductible under Section 80TTB.`,
    why: `Your FD interest this year is ₹${fmt(fdInterest)}. You can claim ₹${fmt(eligible)} as a deduction under 80TTB, available exclusively to senior citizens. This reduces your taxable income directly.`,
    actionLabel: 'Claim 80TTB Deduction',
    actionUrl: '/tax/80ttb',
    authLevel: 2,
    disclosure: TAX_DISCLAIMER,
    dataUsed: ['TAX_PROFILE', 'TRANSACTION_HISTORY'],
    explainRef: `/why/${cardId}`,
    confidence: 0.95,
    isTimeSensitive: false,
  };
}

// ── Card Renderer ──────────────────────────────────────────────────────
export function renderCard(card, parentEl) {
  const div = document.createElement('div');
  div.className = 'rec-card';
  div.id = card.cardId;
  div.innerHTML = `
    <div class="rec-card-accent ${card.accent || ''}"></div>
    <div class="rec-card-body">
      <div class="rec-type-badge">${card.typeBadge}</div>
      <div class="rec-what">${card.what}</div>
      <div class="rec-why">${card.why}</div>
      <div class="rec-action-row">
        <button class="rec-cta-btn ${card.accent === 'fraud' ? 'danger' : ''}"
          onclick="handleCardAction('${card.cardId}', '${card.actionUrl}', ${card.authLevel})">
          ${card.actionLabel} →
        </button>
        <a href="${card.explainRef}" class="rec-why-link">Why am I seeing this?</a>
      </div>
      <div class="confidence-bar">
        <span class="confidence-label">Confidence</span>
        <div class="confidence-track">
          <div class="confidence-fill" style="width:${(card.confidence * 100).toFixed(0)}%"></div>
        </div>
        <span class="confidence-label">${(card.confidence * 100).toFixed(0)}%</span>
      </div>
      <div class="sebi-disclosure">${card.disclosure}</div>
    </div>
  `;
  parentEl.appendChild(div);
  return div;
}

export function renderCardGrid(cards, gridEl) {
  gridEl.innerHTML = '';
  cards.forEach(card => renderCard(card, gridEl));
}

// ── Conversational Avatar Responses ───────────────────────────────────
export function getAvatarResponse(userMessage, segment, context) {
  const msg = userMessage.toLowerCase();
  const persona = AVATAR_PERSONAS[segment] || AVATAR_PERSONAS.YOUNG_EMPLOYED;
  const tone = SEGMENT_TONES[segment];

  if (msg.includes('score') || msg.includes('whs') || msg.includes('health')) {
    return `Your Wealth Health Score is **${context.whs || 72}/100** (Grade ${context.whsGrade || 'B'}). ${context.topAction ? `Your biggest opportunity right now: **${context.topAction}**.` : ''}`;
  }
  if (msg.includes('tax') || msg.includes('80c') || msg.includes('save')) {
    return `You have **₹${fmt(context.sec80cRemaining || 45000)}** unused in your 80C limit this financial year. An ELSS SIP started today could save you up to **₹${fmt(Math.round((context.sec80cRemaining || 45000) * 0.30))}** in tax by March 31. Want me to show you the best options?`;
  }
  if (msg.includes('portfolio') || msg.includes('invest') || msg.includes('fund')) {
    return `Your portfolio is ${context.driftAlerts > 0 ? `showing **${context.driftAlerts} drift alert${context.driftAlerts > 1 ? 's' : ''}**` : 'well-balanced'}. ${context.driftAlerts ? 'I recommend reviewing your asset allocation.' : 'Keep your SIPs running consistently!'}`;
  }
  if (msg.includes('fraud') || msg.includes('safe') || msg.includes('secure')) {
    return `Your account security is ${context.fraudAlerts ? 'Flagged — please review the alert on your dashboard immediately' : 'Normal. No suspicious activity detected in the last 30 days'}.`;
  }
  if (msg.includes('hello') || msg.includes('hi') || msg.includes('help')) {
    return `${persona.greeting} I'm ${persona.name}, your AI wealth advisor. Ask me about your Wealth Health Score, tax savings, portfolio, or anything financial!`;
  }

  return `I can help with your Wealth Health Score, tax optimization (80C, 80D, NPS), portfolio rebalancing, and financial planning. What would you like to explore first?`;
}

export { SEBI_FOOTER, TAX_DISCLAIMER, AVATAR_PERSONAS };
